src/eos_gates.js

```diff
@@ -143,6 +245,162 @@ const Application = new Lang.Class({
+const EVENT_LAUNCHED_EQUIVALENT_EXISTING_FLATPAK = '00d7bc1e-ec93-4c53-ae78-a6b40450be4a';
+
+function installAppFromStore(replacement, originalPayload) {
+    let appStoreId = getAppStoreAppId(replacement.flatpakInfo.remote,
+                                      replacement.flatpakInfo.id);
+    recordMetrics(replacement.replacementInfo ?
+                  EVENT_LAUNCHED_EQUIVALENT_INSTALLER_FOR_FLATPAK :
+                  EVENT_LAUNCHED_INSTALLER_FOR_FLATPAK,
+                  new GLib.Variant('(sas)', [replacement.flatpakInfo.id, originalPayload]));
+    Gio.DBusActionGroup.get(Gio.Application.get_default().get_dbus_connection(),
+                            'org.gnome.Software',
+                            '/org/gnome/Software')
+                       .activate_action('details',
+                                        new GLib.Variant('(ss)', [appStoreId, '']));
+}
+
+function launchFlatpakAppForInstallation(installation, replacement) {
+    let ref = null;
+
+    try {
+        ref = installation.get_current_installed_app(replacement.flatpakInfo.id, null);
+    } catch (e) {
+        return false;
+    }
+
+    installation.launch(replacement.flatpakInfo.id,
+                        null,
+                        'master',
+                        ref.get_latest_commit(),
+                        null);
+
+    return true;
+}
+
+function launchFlatpakApp(replacement, originalPayload) {
+    try {
+        recordMetrics(replacement.replacementInfo ?
+                      EVENT_LAUNCHED_EQUIVALENT_EXISTING_FLATPAK :
+                      EVENT_LAUNCHED_EXISTING_FLATPAK,
+                      new GLib.Variant('(sas)', [replacement.flatpakInfo.id, originalPayload]));
+        if (!launchFlatpakAppForInstallation(Flatpak.Installation.new_user(null), replacement))
+            launchFlatpakAppForInstallation(Flatpak.Installation.new_system(null), replacement);
+    } catch (e) {
+        logError(e, 'Something went wrong in launching %s'.format(replacement.flatpakInfo.id));
+    }
+}
+
+function launchDesktopApp(replacement, originalPayload) {
+    try {
+        recordMetrics(replacement.replacementInfo ?
+                      EVENT_LAUNCHED_EQUIVALENT_EXISTING_FLATPAK :
+                      EVENT_LAUNCHED_EXISTING_FLATPAK,
+                      new GLib.Variant('(sas)', [replacement.desktopInfo.get_id(), originalPayload]));
+        replacement.desktopInfo.launch([], null);
+    } catch (e) {
+        logError(e, 'Something went wrong in launching %s'.format(replacement.desktopInfo.id));
+    }
+}
+
+function launchLink(replacement, originalPayload) {
+    try {
+        recordMetrics(replacement.replacementInfo ?
+                      EVENT_LAUNCHED_EQUIVALENT_EXISTING_FLATPAK :
+                      EVENT_LAUNCHED_EXISTING_FLATPAK,
+                      new GLib.Variant('(sas)', [replacement.linkInfo.href, originalPayload]));
+        let path = Gio.File.new_for_uri(replacement.linkInfo.href);
```

Comment:
I'd just use `Gio.AppInfo.launch_default_for_uri()` -- or even better `Gtk.show_uri()` for this.